### Problema

Para empezar la empresa quiere investigar las diferencias clave entre los empleados que han dejado la organización y los que permanecen en ella. Para ello, nos han pedido que realicemos una serie de análisis los cuales tenemos descritos a continuación: 

1. ¿Existe una diferencia en la edad promedio entre los empleados que han dejado la empresa y los que no?

2. ¿Existe una diferencia en el ingreso mensual promedio entre empleados que han dejado la empresa y los que no?

3. ¿Existe una diferencia en la proporción de empleados que hacen horas extras entre los que han dejado la empresa y los que no?

4. ¿Existe una diferencia en los años promedio trabajados con el mismo gerente entre empleados que han dejado la empresa y los que no?

5. ¿Existe una diferencia en los años desde la última promoción entre los empleados que han dejado la empresa y los que no?

Basándonos en los resultados de los análisis anteriores, escribe aquellos factores que parecen estar más asociados con los empleados que han dejado la empresa en comparación con los que se han quedado.

Llegados a este punto, ya sabemos y hemos identificado algunos patrones importantes entre los empleados que han dejado la empresa y los que no. Investiguemos ahora que esta ocurriendo entre los empleados que han dejado la empresa a traves de las siguientes preguntas: 

- Analiza a los empleados que tienen más de 10 años de experiencia en la empresa y examina cómo varía su tasa de incremento salarial y sus horas estándar de trabajo según su nivel de satisfacción laboral y estado civil.

- Ahora deberéis focalizaros en los empleados casados que tienen una alta satisfacción con su entorno laboral y que pertenecen al departamento de `'Research & Development'`. Deberéis evaluar cómo varían el tiempo desde su última promoción y los años que han estado en su rol actual según su género.

In [4]:
import pandas as pd
import numpy as np

In [5]:
df = pd.read_csv("datos/IBM_HR_Employee_Attrition_full.csv")
pd.set_option("display.max_columns", None)
df.head(10)

Unnamed: 0  DistanceFromHome      Education    EducationField  Gender  \
0           0                21       Bachelor             Other    Male   
1           1                25        College  Technical Degree  Female   
2           2                 1        College     Life Sciences    Male   
3           3                 4        College  Technical Degree    Male   
4           4                 2         Master     Life Sciences  Female   
5           5                10       Bachelor           Medical    Male   
6           6                 6       Bachelor           Medical  Female   
7           7                28  Below College           Medical    Male   
8           8                 6       Bachelor           Medical  Female   
9           9                 9         Master     Life Sciences    Male   

  MaritalStatus  Age                            EmployeeId  \
0        Single   19  70b8db40-4f58-476f-8776-c00802b0cdb9   
1       Married   30  0722da7a-530b-41dd-b705-c79f7627f677   
2        Single   22  b5fed13f-7785-4daf-bb8c-1037f84e5fa7   
3       Married   42  4b0a1169-b286-44fb-a453-b97ca934db57   
4        Single   36  6affb754-8e52-43a7-9e97-4e373ef5fccb   
5        Single   32  098c5463-e599-44ad-95f7-922ffe4b3d20   
6       Married   43  69658f26-48fa-4573-8317-2668b70f54a9   
7       Married   50  e3b6a3f7-9046-494f-a3a3-396caa04c5b8   
8      Divorced   38  0ce9edf2-7655-47ab-918d-2e6f511ace72   
9       Married   41  4c27df95-d97d-4165-8792-f39a3477a650   

        DateEmployment  YearsInCurrentRole  YearsSinceLastPromotion  \
0  2000-01-13 02:53:47                  14                        0   
1  2000-01-17 03:53:10                  16                        0   
2  2000-01-21 05:42:21                  20                        0   
3  2000-01-28 16:59:19                  11                        1   
4  2000-02-02 22:58:25                  13                        0   
5  2000-02-06 14:19:30                  15                        0   
6  2000-02-14 15:46:42                   4                        0   
7  2000-02-19 14:51:34                   2                        0   
8  2000-02-22 11:00:45                   4                        0   
9  2000-03-04 21:27:50                  23                        1   

   YearsWithCurrManager  TotalWorkingYears Over18  NumCompaniesWorked  \
0                     0                  1    NaN                 1.0   
1                     0                  4      Y                 7.0   
2                     0                  4      Y                 0.0   
3                     5                 22    NaN                 1.0   
4                     7                 17    NaN                 3.0   
5                     0                 10    NaN                 2.0   
6                     6                 21    NaN                 7.0   
7                     2                 20      Y                 1.0   
8                     8                 17    NaN                 7.0   
9                     1                  7      Y                 2.0   

           Country  Sons Attrition BusinessTravel  DailyRate  \
0          Austria     1       Yes  Travel_Rarely      419.0   
1           España     1        No     Non-Travel      641.0   
2  República Checa     1        No  Travel_Rarely      581.0   
3         Misisipi     1        No  Travel_Rarely      532.0   
4         Portugal     1        No  Travel_Rarely      530.0   
5          Noruega     0        No  Travel_Rarely      495.0   
6           Italia     0        No  Travel_Rarely        NaN   
7          Bélgica     1        No  Travel_Rarely     1207.0   
8          Croacia     1        No     Non-Travel      573.0   
9           Grecia     0        No  Travel_Rarely      933.0   

               Department  EmployeeCount  HourlyRate  JobLevel  \
0                   Sales              1          37         1   
1                   Sales              1          85        

In [6]:
filtro1 = df["Attrition"] == "Yes"
filtro2 = df["Attrition"] == "yes"

df_desempleado = df[filtro1|filtro2]
df_desempleado

Unnamed: 0  DistanceFromHome      Education    EducationField  Gender  \
0              0                21       Bachelor             Other    Male   
11            11                 4  Below College  Technical Degree    Male   
21            21                 3        College           Medical  Female   
22            22                 2         Master     Life Sciences    Male   
25            25                24        College  Technical Degree    Male   
...          ...               ...            ...               ...     ...   
1422        1422                16         Master     Life Sciences  Female   
1430        1430                 2       Bachelor     Life Sciences  Female   
1444        1444                20       Bachelor     Life Sciences  Female   
1455        1455                 9        College           Medical    Male   
1457        1457                 4         Master         Marketing    Male   

     MaritalStatus  Age                            EmployeeId  \
0           Single   19  70b8db40-4f58-476f-8776-c00802b0cdb9   
11          Single   22  3c212f34-4095-470c-b8a7-1147e4fc38bb   
21          Single   18  6915942d-c6f3-46b6-9858-6017853ca41c   
22          Single   58  326d3298-faf1-4ce9-a5e0-66169bb095cb   
25          Single   28  f3442d9f-b942-42ad-8d87-6f8d4136c634   
...            ...  ...                                   ...   
1422        Single   36  b9c9fcd5-f367-4d5c-b411-16377f734ec3   
1430        Single   39  7fa9e1bc-23dc-4776-a8db-6c8ebb366d75   
1444       Married   31  8dfd9fe9-d3df-4466-a2ef-9c8b4bf8a6aa   
1455        Single   46  f2f066bd-6faf-41a1-bd10-d05a12de5126   
1457        Single   26  75ca9b8b-7b09-4c50-8870-4593b48dbcaa   

           DateEmployment  YearsInCurrentRole  YearsSinceLastPromotion  \
0     2000-01-13 02:53:47                  14                        0   
11    2000-03-20 05:07:49                  13                        1   
21    2000-04-28 12:22:17                   0                        0   
22    2000-06-07 14:22:22                  13                        9   
25    2000-06-16 02:02:31                  20                        0   
...                   ...                 ...                      ...   
1422  2022-03-03 18:34:46                   1                        9   
1430  2022-04-25 18:43:27                   0                        0   
1444  2022-07-07 05:10:24                   0                        0   
1455  2022-08-31 08:57:45                   0                        0   
1457  2022-09-14 07:44:23                   0                        0   

      YearsWithCurrManager  TotalWorkingYears Over18  NumCompaniesWorked  \
0                        0                  1    NaN                 1.0   
11                       2                  4    NaN                 5.0   
21                       0                  0      Y                 1.0   
22                       1                 40    NaN                 7.0   
25                       4                 10      Y                 1.0   
...                    ...                ...    ...                 ...   
1422                     5                 18      Y                 1.0   
1430                     0                 11      Y                 2.0   
1444                     1                  4      Y                 3.0   
1455                     0                  9      Y                 1.0   
1457                     0                  8      Y                 1.0   

      Country  Sons Attrition     BusinessTravel  DailyRate  \
0     Austria     1       Yes      Travel_Rarely      419.0   
11    Noruega     0       Yes  Travel_Frequently     1368.0   
21    Indiana     1       Yes  Travel_Frequently      544.0   
22      Maine     1       Yes      Travel_Rarely      286.0   
25     Alaska     0       Yes                NaN     1366.0   
...       ...   ...       ...                ...        ...   
1422    Suiza     0       Yes    

In [7]:
filtro1 = df["Attrition"] == "No"
filtro2 = df["Attrition"] == "no"

df_empleado = df[filtro1|filtro2]
df_empleado

Unnamed: 0  DistanceFromHome Education    EducationField  Gender  \
1              1                25   College  Technical Degree  Female   
2              2                 1   College     Life Sciences    Male   
3              3                 4   College  Technical Degree    Male   
4              4                 2    Master     Life Sciences  Female   
5              5                10  Bachelor           Medical    Male   
...          ...               ...       ...               ...     ...   
1465        1465                 1   College           Medical  Female   
1466        1466                 7   College           Medical    Male   
1467        1467                 4       NaN           Medical  Female   
1468        1468                 2       NaN   Human Resources  Female   
1469        1469                 5    Master  Technical Degree    Male   

     MaritalStatus  Age                            EmployeeId  \
1          Married   30  0722da7a-530b-41dd-b705-c79f7627f677   
2           Single   22  b5fed13f-7785-4daf-bb8c-1037f84e5fa7   
3          Married   42  4b0a1169-b286-44fb-a453-b97ca934db57   
4           Single   36  6affb754-8e52-43a7-9e97-4e373ef5fccb   
5           Single   32  098c5463-e599-44ad-95f7-922ffe4b3d20   
...            ...  ...                                   ...   
1465        Single   30  ee51e04a-ae21-423f-91ae-bd85c93ed69d   
1466       Married   39  7229488d-28e5-4e95-8df4-d83c5c110722   
1467       Married   24  d612a787-00ab-4769-a965-2b4a625b27d9   
1468       Married   25  7261478d-ea6b-4588-b803-5807218f3a00   
1469       Married   31  34320d20-cf61-4900-9450-c8539f23b355   

           DateEmployment  YearsInCurrentRole  YearsSinceLastPromotion  \
1     2000-01-17 03:53:10                  16                        0   
2     2000-01-21 05:42:21                  20                        0   
3     2000-01-28 16:59:19                  11                        1   
4     2000-02-02 22:58:25                  13                        0   
5     2000-02-06 14:19:30                  15                        0   
...                   ...                 ...                      ...   
1465  2022-12-11 07:01:42                   2                        3   
1466  2022-12-11 16:04:20                   2                       13   
1467  2022-12-14 21:58:45                   2                        0   
1468  2022-12-27 21:53:23                   0                        1   
1469  2022-12-30 07:15:12                   2                        0   

      YearsWithCurrManager  TotalWorkingYears Over18  NumCompaniesWorked  \
1                        0                  4      Y                 7.0   
2                        0                  4      Y                 0.0   
3                        5                 22    NaN                 1.0   
4                        7                 17    NaN                 3.0   
5                        0                 10    NaN                 2.0   
...                    ...                ...    ...                 ...   
1465                     6                 12      Y                 1.0   
1466                     0                 21      Y                 1.0   
1467                     0                  5      Y                 1.0   
1468                     2                  6      Y                 4.0   
1469                     3                  6      Y                 0.0   

              Country  Sons Attrition     BusinessTravel  DailyRate  \
1              España     1        No         Non-Travel      641.0   
2     República Checa     1        No      Travel_Rarely      581.0   
3            Misisipi     1        No      Travel_Rarely      532.0   
4            Portugal     1        No      Travel_Rarely      530.0   
5             Noruega     0        No      Travel_Rarely      495.0   
...               ...   ...       ...                ...        ...   
1465          Noruega     0        No

1. ¿Existe una diferencia en la edad promedio entre los empleados que han dejado la empresa y los que no?

In [8]:
df_empleado.describe().T

count          mean          std     min      25%  \
Unnamed: 0               1233.0    732.562044   424.681016     1.0   365.00   
DistanceFromHome         1233.0      8.915653     8.012633     1.0     2.00   
Age                      1233.0     37.561233     8.888360    18.0    31.00   
YearsInCurrentRole       1233.0      6.276561     5.371332     0.0     2.00   
YearsSinceLastPromotion  1233.0      1.063260     1.947164     0.0     0.00   
YearsWithCurrManager     1233.0      2.218978     2.688491     0.0     0.00   
TotalWorkingYears        1233.0     11.862936     7.760719     0.0     6.00   
NumCompaniesWorked       1221.0      2.639640     2.457748     0.0     1.00   
Sons                     1233.0      0.509327     0.500116     0.0     0.00   
DailyRate                1114.0    814.614901   405.924486   102.0   474.75   
EmployeeCount            1233.0      1.000000     0.000000     1.0     1.00   
HourlyRate               1233.0     65.952149    20.380754    30.0    48.00   
JobLevel                 1233.0      2.145985     1.117933     1.0     1.00   
MonthlyIncome            1233.0   6832.739659  4818.208001  1051.0  3211.00   
MonthlyRate              1233.0  14265.779400  7102.260749  2094.0  7973.00   
PercentSalaryHike        1233.0     15.231144     3.639511    11.0    12.00   
StandardHours             980.0     80.000000     0.000000    80.0    80.00   
StockOptionLevel         1233.0      0.845093     0.841985     0.0     0.00   
TrainingTimesLastYear    1233.0      2.832928     1.293585     0.0     2.00   

                             50%       75%      max  
Unnamed: 0                 730.0   1093.00   1469.0  
DistanceFromHome             7.0     13.00     29.0  
Age                         36.0     43.00     60.0  
YearsInCurrentRole           5.0     10.00     23.0  
YearsSinceLastPromotion      0.0      1.00     13.0  
YearsWithCurrManager         1.0      3.00     17.0  
TotalWorkingYears           10.0     16.00     38.0  
NumCompaniesWorked           2.0      4.00      9.0  
Sons                         1.0      1.00      1.0  
DailyRate                  812.5   1180.75   1499.0  
EmployeeCount                1.0      1.00      1.0  
HourlyRate                  66.0     83.00    100.0  
JobLevel                     2.0      3.00      5.0  
MonthlyIncome             5204.0   8834.00  19999.0  
MonthlyRate              14120.0  20364.00  26997.0  
PercentSalaryHike           14.0     18.00     25.0  
StandardHours               80.0     80.00     80.0  
StockOptionLevel             1.0      1.00      3.0  
TrainingTimesLastYear        3.0      3.00      6.0

In [9]:
df_desempleado.describe().T

count          mean          std     min     25%  \
Unnamed: 0               237.0    744.582278   424.290633     0.0   391.0   
DistanceFromHome         237.0     10.632911     8.452525     1.0     3.0   
Age                      237.0     33.607595     9.689350    18.0    28.0   
YearsInCurrentRole       237.0      6.025316     5.344825     0.0     2.0   
YearsSinceLastPromotion  237.0      0.936709     1.882176     0.0     0.0   
YearsWithCurrManager     237.0      1.379747     2.078770     0.0     0.0   
TotalWorkingYears        237.0      8.244726     7.169204     0.0     3.0   
NumCompaniesWorked       233.0      2.871245     2.639339     0.0     1.0   
Sons                     237.0      0.468354     0.500054     0.0     0.0   
DailyRate                209.0    740.315789   401.351414   103.0   383.0   
EmployeeCount            237.0      1.000000     0.000000     1.0     1.0   
HourlyRate               237.0     65.573840    20.099958    31.0    50.0   
JobLevel                 237.0      1.637131     0.940594     1.0     1.0   
MonthlyIncome            237.0   4787.092827  3640.210367  1009.0  2373.0   
MonthlyRate              237.0  14559.308017  7208.153264  2326.0  8870.0   
PercentSalaryHike        237.0     15.097046     3.770294    11.0    12.0   
StandardHours            211.0     80.000000     0.000000    80.0    80.0   
StockOptionLevel         237.0      0.527426     0.856361     0.0     0.0   
TrainingTimesLastYear    237.0      2.624473     1.254784     0.0     2.0   

                             50%      75%      max  
Unnamed: 0                 751.0   1127.0   1457.0  
DistanceFromHome             9.0     17.0     29.0  
Age                         32.0     39.0     58.0  
YearsInCurrentRole           4.0      9.0     23.0  
YearsSinceLastPromotion      0.0      1.0     11.0  
YearsWithCurrManager         0.0      2.0     11.0  
TotalWorkingYears            7.0     10.0     40.0  
NumCompaniesWorked           1.0      5.0      9.0  
Sons                         0.0      1.0      1.0  
DailyRate                  688.0   1089.0   1496.0  
EmployeeCount                1.0      1.0      1.0  
HourlyRate                  66.0     84.0    100.0  
JobLevel                     1.0      2.0      5.0  
MonthlyIncome             3202.0   5916.0  19859.0  
MonthlyRate              14618.0  21081.0  26999.0  
PercentSalaryHike           14.0     17.0     25.0  
StandardHours               80.0     80.0     80.0  
StockOptionLevel             0.0      1.0      3.0  
TrainingTimesLastYear        2.0      3.0      6.0

Edad media de los empleados = 37.5

Edad media de los que han dejado la empresa = 33.6

La mediana y la media de ambos datos son muy próximas.


2. ¿Existe una diferencia en el ingreso mensual promedio entre empleados que han dejado la empresa y los que no?

Sueldo mensual promedio de los empleados = 6832.7 y la mediana 5204

Sueldo mensual promedio de los que han dejado la empresa = 4787.1 y la mediana 3202.


3. ¿Existe una diferencia en la proporción de empleados que hacen horas extras entre los que han dejado la empresa y los que no?

In [10]:
df_empleado["OverTime"].value_counts()/df_empleado.shape[0]*100

OverTime
No     76.561233
Yes    23.438767
Name: count, dtype: float64

In [11]:
df_desempleado["OverTime"].value_counts()/df_desempleado.shape[0]*100

OverTime
Yes    53.586498
No     46.413502
Name: count, dtype: float64

De los empleados el 23.4% hacen horas extra.

De los que han dejado la empresa el 53.6% hacían horas extra.

4. ¿Existe una diferencia en los años promedio trabajados con el mismo gerente entre empleados que han dejado la empresa y los que no?

Los empleados han estado una media de 2.2 años con su actual manager.

Los que han dejado la empresa han estado una media de 1.4 años con su último manager.

5. ¿Existe una diferencia en los años desde la última promoción entre los empleados que han dejado la empresa y los que no?

Los empleados tienen una media de 1.1 años desde la ultima promocion.

Los que han dejado la empresa tenían una media de 0.9 años desde la ultima promocion.

In [12]:
filtro = df_desempleado["TotalWorkingYears"] > 10
df_desempleado_10 = df_desempleado[filtro]
df_desempleado_10.groupby(["JobSatisfaction", "MaritalStatus"]).mean(numeric_only=True)[["StandardHours", "PercentSalaryHike"]]

StandardHours  PercentSalaryHike
JobSatisfaction MaritalStatus                                  
High            Divorced                80.0          16.500000
                Married                 80.0          13.222222
                Single                  80.0          15.000000
Low             Divorced                80.0          13.666667
                Married                 80.0          16.833333
                Single                  80.0          16.375000
Medium          Married                 80.0          18.250000
                Single                  80.0          17.000000
Very High       Divorced                80.0          11.000000
                Married                 80.0          13.250000
                Single                  80.0          18.333333

In [13]:
filtro = df_desempleado["MaritalStatus"] == "Married"
filtro1 = df_desempleado["EnvironmentSatisfaction"] == "High"
filtro2 = df_desempleado["Department"] == "Research & Development"
df_desempleado_casado_reach = df_desempleado[filtro & filtro1 & filtro2]
df_desempleado_casado_reach.groupby("Gender").mean(numeric_only=True)[["YearsSinceLastPromotion", "YearsInCurrentRole"]]

YearsSinceLastPromotion  YearsInCurrentRole
Gender                                             
Female                    0.500            3.666667
Male                      1.125            5.250000

In [14]:
df_desempleado_casado_reach.groupby("Gender").median(numeric_only=True)[["YearsSinceLastPromotion", "YearsInCurrentRole"]]

YearsSinceLastPromotion  YearsInCurrentRole
Gender                                             
Female                      0.0                 2.5
Male                        0.0                 2.0

***Facts:***  
  
Hemos comprobado que los empleados que abandonaron la empresa en media son más jóvenes (33,6 frente a 37,5), cobraban menos (4787,1 frente a 6832,7) pero ambos tienen datos atípicos al alza (Es decir, alguna persona que cobra mucho y descuadra la media), muchos de ellos hacían horas extras (53,6% frente al 23,4% de los empleados actuales), han estado menos años con el mismo manager(1,4 años frente a 2,2 años) y han sido promoconados en menos de un año frente al año de un empleado actual.

Además los que abandonaron 

**Conclusiones:**  

Estos datos nos pueden indicar que los empleados que abandonaron la empresa pudieron tener como aliciente las malas condiciones salariales y de cumplimiento de horas pactadas en el contrato. Además, la empresa parece haberse centrado en promocionar más rápido a estos disertantes (esto a su vez explica porque están menos años en media con el mismo manager) lo que podría estar desequilibrando el ratio salario-responsabilidades lo que podría ser otra razón. Como posible solución IBM podría restructurar sus sistemas de recompensas y que cada promoción venga acompañada de un equilibrio entre el aumento de carga de trabajo y el aumento salarial.




Sobre los empleados que llevan más de 10 años en la empresa, se observa como los solteros cuentan con  

